base-containers/base/bin/_run_student_intern

```diff
@@ -5,80 +5,43 @@
+    run_teardown_script, handle_signals, handle_ssh_session, receive_initial_command, stdio,\
+    handle_stdin, handle_outputs_helper
 
 
-def handle_signals(subprocess, socket):
-    """ Handles signals given by run_student on the socket """
-    while True:
-        signal = socket.recv(3)
-        if signal == b'---' or len(signal) < 3: # quit
-            return
-        print("received %s" % signal.decode("utf8"))
-        subprocess.send_signal(int(signal.decode('utf8')))
+# Setup the logger
+logger = setup_logger()
 
-
-def recv_fds(sock, msglen, maxfds):
-    """ Receive FDs from the unix socket. Copy-pasted from the Python doc """
-    fds = array.array("i")  # Array of ints
-    msg, ancdata, flags, addr = sock.recvmsg(msglen, socket.CMSG_LEN(maxfds * fds.itemsize))
-    for cmsg_level, cmsg_type, cmsg_data in ancdata:
-        if (cmsg_level == socket.SOL_SOCKET and cmsg_type == socket.SCM_RIGHTS):
-            # Append data, ignoring any truncated integers at the end.
-            fds.fromstring(cmsg_data[:len(cmsg_data) - (len(cmsg_data) % fds.itemsize)])
-    return msg, list(fds)
-
-
-# Check the runtime
+# Check the runtimes
 runtime = sys.argv[1]
-if runtime == "kata-runtime":
-    shared_kernel = False
+parent_runtime = sys.argv[2]
+shared_kernel, only_dockers = check_runtimes(runtime, parent_runtime)
+
+# Setup the asyncio loop and container_stdin/stdout streams for communications with agent when both containers are not on a shared kernel
+if not only_dockers:
+    context = zmq.asyncio.Context()
+    event_loop = zmq.asyncio.ZMQEventLoop()
+    asyncio.set_event_loop(event_loop)
+    container_stdin, container_stdout = event_loop.run_until_complete(stdio())
 else:
-    shared_kernel = True
-    os.mkdir("/.__input")
-    shared_kernel_file = open("/.__input/__shared_kernel", "w")
-    shared_kernel_file.close()
-
+    container_stdin, container_stdout, event_loop = None, None, None
 
-# Connect to the socket
-client = socket.socket(socket.AF_UNIX)  # , socket.SOCK_CLOEXEC) # for linux only
-client.connect("/__parent.sock")
 
-# Say hello
-print("Saying hello")
-client.send(b'H')
-print("Said hello")
+# Get the command to be run
+socket_unix, fds, start_cmd = receive_initial_command(only_dockers, container_stdin, event_loop)
 
-# Receive fds
-print("Receiving fds")
-msg, fds = recv_fds(client, 1, 3)
-assert msg == b'S'
-print("Received fds")
-
-# Unpack the start message # TODO: it's ugly
-print("Unpacking start cmd")
-unpacker = msgpack.Unpacker()
-start_cmd = None
-while start_cmd is None:
-    s = client.recv(1)
-    unpacker.feed(s)
-    for obj in unpacker:
-        start_cmd = obj
-print("Unpacked start cmd")
+# Check security
+student_container_id = start_cmd["student_container_id"]
 user = start_cmd["user"]  # Know if you are worker (on docker) or root (on Kata)
-
 if user == "root" and shared_kernel:
-    print("You can not run as root on docker runtime")
```

Comment:
Added a logger message.
Note using "print" in the run_student_intern script itself is prohibited now since the print would be received by the agent with an incorrect format.